### ISSUE TITLE
Filtered search does not take recently updated document into account

### ISSUE THREAD

**Issue by daFish** [meilisearch/meilisearch#3887]
**Describe the bug**
When I update a document already in the index, a subsequent search with a filtering will not return the just-updated document. Fetch the document itself from the documents-endpoint returns the document. I also get the document from the search when I do not use any filters.

When I re-index all documents the issue is fixed. But this is not a desired workaround.

*Note* The filter in question does not filter out the aforementioned document. New documents are not affected by this issue.

**To Reproduce**
1. Update document
2. Search with filter (which will return said document)
3. Document is not returned by search

**Expected behavior**
The updated document is returned from search with or without any filter.

**Meilisearch version:**
1.2.0 (`getmeili/meilisearch:v1.2.0`)

**Additional context**
* ~7500 documents in the index
* no pending tasks
* Docker 24.0.2
* PHP 8.2.7
* meilisearch/meilisearch-php (1.2.1)
* meilisearch/search-bundle (0.13.0)

**Comment by gillian-meilisearch** [comment]
Hi @daFish thanks so much for reporting this 🙏  Can I ask is this something you have experienced in previous versions of Meilisearch such as v1.1 as well, or do you feel this is newly introduced on v1.2.0?
If you're able to provide more details/examples on your steps to reproduce that would also be super helpful for us to look into this further. In the meantime we'll see what we can find here. Thanks!

**Comment by daFish** [comment]
Hi @gillian-meilisearch,

unfortunatly I have not encountered this issue with 1.1.

I'll try to reproduce this with a demo application but in the meantime here are the latest entries from the task queue:
```{
            "uid": 2436,
            "indexUid": "dev_releases",
            "status": "succeeded",
            "type": "documentAdditionOrUpdate",
            "canceledBy": null,
            "details": {
                "receivedDocuments": 1,
                "indexedDocuments": 1
            },
            "error": null,
            "duration": "PT0.097757504S",
            "enqueuedAt": "2023-07-05T05:54:34.033915771Z",
            "startedAt": "2023-07-05T05:54:34.03539343Z",
            "finishedAt": "2023-07-05T05:54:34.133150934Z"
        },
        {
            "uid": 2435,
            "indexUid": "dev_releases",
            "status": "succeeded",
            "type": "documentAdditionOrUpdate",
            "canceledBy": null,
            "details": {
                "receivedDocuments": 1,
                "indexedDocuments": 1
            },
            "error": null,
            "duration": "PT0.139538317S",
            "enqueuedAt": "2023-07-05T05:54:13.030244838Z",
            "startedAt": "2023-07-05T05:54:13.031794579Z",
            "finishedAt": "2023-07-05T05:54:13.171332896Z"
        },
        {
            "uid": 2434,
            "indexUid": "dev_releases",
            "status": "succeeded",
            "type": "documentAdditionOrUpdate",
            "canceledBy": null,
            "details": {
                "receivedDocuments": 1,
                "indexedDocuments": 1
            },
            "error": null,
            "duration": "PT0.248491561S",
            "enqueuedAt": "2023-07-05T05:52:46.35382957Z",
            "startedAt": "2023-07-05T05:52:46.355492353Z",
            "finishedAt": "2023-07-05T05:52:46.603983914Z"
        }
```

**Comment by gillian-meilisearch** [comment]
Hi @daFish, thanks for that information, they seem pretty quick so likely not an issue.

I think perhaps the issue you are seeing is to do with caching of requests on the front end. Please let me know if this doesn't match what you're seeing, but it sounds like it might be.

So for example in the mini-dashboard I search in my movies documents for 'batman'. In the fetch requests you'll see a request for every letter I type (6 sent to /search in total)

![image](https://github.com/meilisearch/meilisearch/assets/132390895/36a1e241-87b5-47f9-b0b1-6efa8c811531)

As I do that I get a batman document back in my results as expected. Great.

If I then, in the background, do an update to that document and, without refreshing my search page, type the same letters again 'batman' you'll see there are no further fetch requests made. That's because there's a level of caching on the front end (the exact same request to /search is made so it's not sent again). As there are no new requests you're still seeing the outdated document.

However if I update that search query by anything else other than exactly 'batman' again e.g. I add a space or additional words then I would see my updated document appear. Equally, if I refresh the search page (without any indexing changes) and do the exact search again I'll also get the updated document.

It also makes sense then that if you send a specific fetch request again manually you see the updated document.

Does that sound like what's happening for you?

**Comment by daFish** [comment]
Hi @gillian-meilisearch,

I could reproduce it with the basic `movies.json`-dataset from the "Getting started" guide. These are my steps:

1. Ingest from the file: 
```curl \
  -X POST 'http://localhost:7700/indexes/movies/documents?primaryKey=id' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer masterKey' \
  --data-binary @movies.json
  ```
2. Update settings to have filterable fields:
```curl \
  -X PATCH 'http://localhost:7700/indexes/movies/settings' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer masterKey' \
  --data-binary '{
    "filterableAttributes": [
      "title",
      "genres"
    ],
    "sortableAttributes": [
      "title",
      "release_date"
    ]
  }'
```
3. Search with filter:
```
curl \
  -X POST 'http://localhost:7700/indexes/movies/search' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer masterKey' \
  --data-binary '{
    "filter": "genres = Action"
  }'
```
This will return several documents incl. "Judgment Night"

4. Update document with `primaryId` 6 which is "Judgment Night".
```
curl \
  -X PUT 'http://localhost:7700/indexes/movies/documents' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer masterKey' \
  --data-binary '[
    {
      "id": 6,
      "title": "Judgment Night 🚀"
    }
  ]'
```
5. Re-run filtered search from step 3. The recently updated document is no longer in the results.
6. Fetch the document directly shows the updated content
```
curl \
  -X GET 'http://localhost:7700/indexes/movies/documents/6?' \
  -H 'Authorization: Bearer masterKey'
```

I might miss something but I'm completely unsure.

**Comment by daFish** [comment]
Thank you very much @gillian-meilisearch.

**Comment by Kerollmops** [comment]
Hey @daFish 👋,

There is no bug here. It's just a matter of ordering and limits. A new internal id is generated when a document is replaced or updated. Those internal ids are incremental. Therefore, the new associated id is no longer a small one but a big one. This means that the document is moved down when you query the database. As the ordering can't be determined between all of the other documents that match the filter, documents are output in internal id order.

When you query the engine with a more restrictive query, i.e.,`"q": "Judgement Night"`, the document is visible. You can also see that the document is present if you add more filters in the settings. You see it because this query much more restricts the number of documents and is output by the engine.

```bash
curl \
  -X POST 'http://localhost:7700/indexes/movies/search' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer masterKey' \
  --data-binary '{
    "filter": "genres = Action",
    "q": "Judgment Night"
  }'

```

You could also argue that you can't see the document by using the first filter query you were using even when you were setting a limit of 10k documents. The reason is that the engine will never return more than 1000 documents by default. That's [the `pagination.maxTotalHits` setting](https://www.meilisearch.com/docs/reference/api/settings#pagination).

**Comment by daFish** [comment]
Hi @Kerollmops,

thanks for the update. The reason with the internal id makes sense. However, I do not understand one thing: How does the internal id affect the sorting for a search? I'll try to reproduce this with.

**Comment by Kerollmops** [comment]
Hey @daFish,

> How does the internal id affect the sorting for a search? I'll try to reproduce this with.

It affects the order of the documents when there is nothing else to determine which one is better than the others. So the order of the documents is undefined when they are considered equal according to the ranking rules.

**Comment by daFish** [comment]
Hello @Kerollmops,

thanks for the clarification. So I won't be able to get my desired result with a query like this?
```
curl \
  --request POST \
  --url 'http://meilisearch:7700/indexes/dev_releases/search' \
  --header 'Authorization: Bearer masterKey' \
  --data '{"q":null,"page":1,"hitsPerPage":20,"facets":["country"],"filter":"labels_ids=\"f9e6ba0a-4ba0-48b8-b97e-bf54a3d4396f\"","sort":["sort_title:asc"]}'
```

**Comment by Kerollmops** [comment]
Hey @daFish 👋

I am not sure that I understand _my desired result_ here. Could you give me more information on what is the desired results? If you want to show the documents in insertion order, you should [sort the documents by the timestamp](https://www.meilisearch.com/docs/learn/core_concepts/relevancy#custom-rules) at which you inserted it in the database.

**Comment by daFish** [comment]
Hi @Kerollmops,

I want to show the documents sorted by another field. In my concrete case sorted by the documents title in ascending order. Once one of the documents is updated I would assume that the order of the results is not changed. However, the just updated document is not returned. This made me create this ticket in the first place because it looks like a strange behaviour to me.

**Comment by Kerollmops** [comment]
Ok, so I understand now. Note that the original order of the documents is non-determined.

If what you want is a stable determined order, I advise you to first sort on the _title_ and just after that on the _id_ maybe 🤔 If this solution doesn't please you, could you please, create a new discussion [on the product repository](https://github.com/meilisearch/product/discussions), explaining your specific use case?

**Comment by daFish** [comment]
Hello @Kerollmops & @gillian-meilisearch,

sorry for bumping this closed issue. Due to recent activity in https://github.com/orgs/meilisearch/discussions/657 I tried to reproduce my issue again. After I was unable to reproduce it with my steps from above, I dug into the source of https://github.com/meilisearch/meilisearch-php and found that document updates are sent via POST instead of PUT.

tl;dr

Using POST instead of PUT for updating an documents leads to the described behaviour.

**Comment by daFish** [comment]
Ok, after reading through the documentation this is inteded behaviour. `POST` replaces the document but `PUT` updates an existing document.
